src/cli/self_update.rs

```diff
@@ -0,0 +1,276 @@
+use std::{
+    env,
+    io::{Seek, SeekFrom, Write},
+};
+
+use flate2::read::GzDecoder;
+use tar::Archive;
+
+use miette::IntoDiagnostic;
+use reqwest::Client;
+use serde::Deserialize;
+
+/// Update pixi to the latest version or a specific version. If the pixi binary is not found in the default location
+/// (e.g. `~/.pixi/bin/pixi`), pixi won't updated to prevent breaking the current installation (Homebrew, etc).
+/// The behaviour can be overridden with the `--force` flag.
+#[derive(Debug, clap::Parser)]
+pub struct Args {
+    /// The desired version (to downgrade or upgrade to). Update to the latest version if not specified.
+    #[clap(long)]
+    version: Option<String>,
+
+    /// Force the update even if the pixi binary is not found in the default location.
+    #[clap(long)]
+    force: bool,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+struct GithubRelease {
+    tag_name: String,
+    name: String,
+    body: String,
+    html_url: String,
+    assets: Vec<GithubReleaseAsset>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GithubReleaseAsset {
+    name: String,
+    browser_download_url: String,
+}
+
+fn user_agent() -> String {
+    format!("pixi {}", env!("CARGO_PKG_VERSION"))
+}
+
+fn default_archive_name() -> Option<String> {
+    if cfg!(target_os = "macos") {
+        if cfg!(target_arch = "x86_64") {
+            Some("pixi-x86_64-apple-darwin.tar.gz".to_string())
+        } else {
+            Some("pixi-aarch64-apple-darwin.tar.gz".to_string())
+        }
+    } else if cfg!(target_os = "windows") && cfg!(target_arch = "x86_64") {
+        Some("pixi-x86_64-pc-windows-msvc.zip".to_string())
+    } else if cfg!(target_os = "linux") {
+        if cfg!(target_arch = "x86_64") {
+            Some("pixi-x86_64-unknown-linux-musl.tar.gz".to_string())
+        } else if cfg!(target_arch = "aarch64") {
+            Some("pixi-aarch64-unknown-linux-musl.tar.gz".to_string())
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+pub async fn execute(args: Args) -> miette::Result<()> {
+    // If args.force is false and pixi is not installed in the default location, stop here.
+    match (args.force, is_pixi_binary_default_location()) {
+        (false, false) => {
+            println!(
```

Comment:
I suggest we go with that logic for now as it will work 99% of the time for the downstream users, and it's always possible to adapt the logic in the future if you get user's reports about edge cases.

Meanwhile, the `--force` flag will always allow people to update.

That being said, I am fine modifying it if you prefer.